features/64-chat-offline-cache/api.md

```diff
@@ -0,0 +1,373 @@
+#### 2
+
+```
+---
+ ^
+ | older cached result
+ v
+---
+(a gap bewtween two sets of result)
+---
+ ^
+ | newly fetched result
+ v
+---
+```
+
+Solution:
+
+- Newly fetched result is appended to the older cached result
+
+**How to fill the gap?**
+
+The utility functions we provide should be able to tell developers where the boundaries of the message list are.
+
+```
+let boundaries = skygearChat.findBoundaries(messagelist)
+
+// if no gap in the message list
+// boundaries == [[messageA, messageZ]]
+//
+// app should load messages newer than messageA and older than messageZ
+
+// else
+// boundaries == [[messageA, messageD], [messageG, messageZ]]
+//
+// app should load messages newer than messageA or messageG, and older than messageD or messageZ
+```
+
+This could be quite complicated, so we may provide scroll view delegate or handler implementation to handle this logic for the developer.
+
+# Changes Required
+
+### Server
+
+Add a `previous_message` to `message` table, to determine if two given messages are subsequent messages.
+
+Deleted messages are also returned in fetch API, but the content of the message would be empty. This is to keep the whole message list complete.
+
+### SDK
+
+- Add cached callback to fetch messages and fetch conversations API.
+
+```
+func fetchConversations(cachedCallback?)
+func fetchMessages(conversation, beforeMessage, limit, order, cachedCallback?)
+func fetchMessages(conversation, afterMessage, limit, order, cachedCallback?)
+```
+
+#### JS
+
+```js
+const successCallback = (conversations, cached = false) => {
+  // display conversations
+};
+
+skygearChat
+  .fetchConversations(successCallback)
+  .then(successCallback, errorCallback);
+
+// no-cache version
+skygearChat
+  .fetchConversation()
+  .then(successCallback, errorCallback);
+```
+
+```js
+let currentMessages = [];
+
+const successCallback = (messages, cached = false) => {
+  currentMessages = skygearChat.merge(currentMessages, messages);
+  // display currentMessages
+};
+
+skygearChat
+  .fetchMessages(conversation, beforeMessage, limit, order, successCallback)
+  .then(successCallback, errorCallback);
+
+// no-cache version
+skygearChat
+  .fetchMessages(conversation, beforeMessage, limit, order)
+  .then(successCallback, errorCallback);
+```
+
+#### Non-JS
+
+```
+skygearChat.fetchConversations(func (conversations, cached = false) {
+  // display conversations
+})
+```
+
+```
+let currentMessages = [];
+
+skygearChat.fetchMessages(func (messages, cached = false) {
+  currentMessages = skygearChat.merge(currentMessages, messages);
+  // display currentMessages
+})
+```
+
+- Add utility functions for handling result returned by chat container
+
+```
+func merge(conversationDeltas: [ConversationDelta], toList: [Conversation]) -> [Conversation]
+func merge(messages: [Message], toList: [Message]) -> [Message]
+func merge(messageDeltas: [MessageDelta], toList: [Message]) -> [Message]
+
+func findBoundaries(ofMessageList: [Message]) -> [(Int, Int)]
```

Comment:
I'm not quite sure if these are really easy to use... maybe consider going "higher level" to provide the TableViewController / CollectionViewController is better... or we could leave this part empty, and try to do it when we apply the offline cache feature on the UIKit and refactor the functions as utility functions with a better use case?

Reply:
I suspect (wild guess!) the `ManagedMessageList` below would solve more common use case.... just guess again. So I think it is your call...

Reply:
The `findBoundaries` is lower level function. I expect `ManagedMessageList` is enough for most common use case. But if `ManagedMessageList` does not suite someone's requirement, they can still use the `findBoundaries` instead of scanning the message list themselves, just more convenient and tell the developers the gap is supposed to be found and handled.

Reply:
`ManagedMessageList` should be using `findBoundaries` internally.

And I am not sure if `findBoundaries` should be private or public yet.

Reply:
I suggest keeping `findBoundaries` internally at this stage yet. We can always release it for the public in future if we find more use cases, but maintain another set not-quite-well-known public API which we might need to maintain backward compatibility in future is a trouble...

(Haha, shall we write somewhere to reminds ourselves, always release a minimal set of API that's enough for the use cases we designed?)